File: 3756. Concatenate Non-Zero Digits and Multiply by Sum II.py
from typing import List
class Solution:
    MOD = 10**9 + 7
    def sumAndMultiply(self, s: str, queries: List[List[int]]) -> List[int]:
        n=len(s)
        pref_sum = [0] * (n + 1)
        pref_num = [0] * (n + 1)
        pref_cnt = [0] * (n + 1)

        # Powers of 10 modulo MOD
        pow10 = [1] * (n + 1)
        for i in range(1, n + 1):
            pow10[i] = (pow10[i - 1] * 10) % self.MOD
        
        # Build prefix arrays
        for i in range(n):
            pref_sum[i + 1] = pref_sum[i]
            pref_num[i + 1] = pref_num[i]
            pref_cnt[i + 1] = pref_cnt[i]

            if s[i] != '0':
                digit = int(s[i])

                pref_sum[i + 1] += digit
                pref_cnt[i + 1] += 1
                pref_num[i + 1] = (pref_num[i] * 10 + digit) % self.MOD
        
        ans = []
        for l, r in queries:
            # Sum of non-zero digits
            digit_sum = pref_sum[r + 1] - pref_sum[l]

            # Count of non-zero digits
            digit_count = pref_cnt[r + 1] - pref_cnt[l]

            # Concatenated number modulo MOD
            number = (
                pref_num[r + 1]
                - (pref_num[l] * pow10[digit_count]) % self.MOD
                + self.MOD
            ) % self.MOD

            ans.append((number * digit_sum) % self.MOD)

        return ans

File: 3756. Concatenate Non-Zero Digits and Multiply by Sum II.cpp
class Solution {
public:
    static const int MOD = 1e9 + 7;
    vector<int> sumAndMultiply(string s, vector<vector<int>>& queries) {
        int n=s.size();

        vector<long long> prefSum(n + 1, 0);
        vector<long long> prefNum(n + 1, 0);
        vector<int> prefCnt(n + 1, 0);

        //precomputing the power of 10
        vector<long long> pow10(n + 1, 1);
        for (int i = 1; i <= n; i++)
            pow10[i] = (pow10[i - 1] * 10) % MOD; 

        for (int i = 0; i < n; i++) {
            prefSum[i + 1] = prefSum[i];
            prefNum[i + 1] = prefNum[i];
            prefCnt[i + 1] = prefCnt[i];

            if (s[i] != '0') {
                int d = s[i] - '0';

                prefSum[i + 1] += d;
                prefCnt[i + 1]++;

                prefNum[i + 1] = (prefNum[i] * 10 + d) % MOD;
            }
        }
        vector<int> ans;

        for (auto &q : queries) {
            int l = q[0];
            int r = q[1];

            long long sumDigits = prefSum[r + 1] - prefSum[l];

            int cntDigits = prefCnt[r + 1] - prefCnt[l];

            long long number =
                (prefNum[r + 1] -
                 prefNum[l] * pow10[cntDigits] % MOD +
                 MOD) % MOD;

            ans.push_back((number * sumDigits) % MOD);
        }

        return ans;   
    }
};